README.md
# PagePark 

I wrote  this simple HTTP server to park domains I've bought but not yet used. 

Then I kept going and added all the features I want to help me manage my own websites, far beyond just parking them. I liked the name so I kept it. Think of it as a nice park where you keep your pages. ;-)

It's written in JavaScript and runs in Node.js.

Each domain is in its own folder. The content for that domain is in the folder. It can serve Markdown docs, or optionally run JavaScript code. And of course HTML, text files, images, movies, etc. 

Yet it's still very simple. Which is the point. ;-)

It's 90 percent of what all web servers do, so if you learn how to run PagePark, you're learning how to run a web server. A real one you can use to host your sites. And it's easy to hack the code if you want to.

### How to

How to get started quickly with PagePark.

0. Install Node.js if you don't already have it running, including the NPM package manager.

1. Download the PagePark folder <a href="https://github.com/scripting/pagepark/archive/master.zip">from</a> the GitHub repository.

2. At the command line enter <code>npm install</code>.

3. To run the server, enter <code>node pagepark.js</code>.

4. Create a sub-folder of the domains folder called localhost. 

5. Using a text editor add a file to the localhost folder called index.md. Put whatever you like in the file. Save it.

6. In a browser, running on the same machine, enter localhost:1339. You should see the text you entered in the previous step. 

If this worked, congratulations -- you just installed a web server. :-)

### Screen shot

Here's a <a href="http://scripting.com/2015/01/04/pageParkFolderScreenShot.png">screen shot</a> of an example PagePark <i>domains</i> folder. 

### How it works

1. PagePark will automatically create a *prefs* sub-folder and a *domains* sub-folder. 

2. Add your web content under domains. Each folder's name is the name of a domain. The contents within the folder are what we serve. <a href="http://scripting.com/2015/01/04/pageParkFolderScreenShot.png">Screen shot</a>.

3. Serves all major media types including audio and video. Files whose names end with .md are passed through the built-in Markdown processor. Files ending with .js are interpreted as scripts only if the associated <a href="docs/config.md">config</a> setting is turned on. The text they return is what we serve. Here's an <a href="https://gist.github.com/scripting/2b5e237a105b6cb96287">example</a> of a script that I have <a href="http://lucky.wtf/badass/butt.js">running</a> on one of my servers.

4. The prefs folder contains a file of settings you can change, prefs.json. These include the port that the server runs on and the name of the index file (see below).

5. stats.json contains information generated by the server including the number of times the server has started, how many hits it's received (all time and today), and hits by domain.

6. mdTemplate.txt is the template we use to serve Markdown text. You can edit this file to provide a common template for all your Markdown documents.

7. If a request comes in for a folder, we scan the folder for a file whose name begins with *index* and serve the first one we find. So the index file can be HTML, Markdown or a script, or any other type PagePark can serve. 

8. If you want to run PagePark from a folder different from the one that contains the app, set the *pageparkFolderPath* environment variable to point to that folder. 

9. There are three special endpoints on all domains: /version, /now and /status that return the version of PagePark that's running, the time on the server and the stats and prefs. 

### File extensions

The extension of a file determines how PagePark serves it.

<table>

<tr>

<td>

.txt

</td>

<td>

The text in the file is returned, the type is text/plain.

</td>

</tr>

<tr>

<td>

.xml

</td>

<td>

The text in the file is returned, the type is text/xml.

</td>

<td>

<a href="http://lucky.wtf/davetwitterfeed.xml">davetwitterfeed.xml</a>

</td>

</tr>

<tr>

<td>

.json

</td>

<td>

The text in the file is returned, the type is application/json.

</td>

<td>

<a href="http://lucky.wtf/wodemo.json">wodemo.json</a>

</td>

</tr>

<tr>

<td>

.png

</td>

<td>

The contents of the file is returned with type image/png.

</td>

<td>

<a href="http://lucky.wtf/tree.png">tree.png</a>

</td>

</tr>

<tr>

<td>

.opml

</td>

<td>

The outline is rendered as an expandable outline, the type returned is text/html.

</td>

<td>

<a href="http://lucky.wtf/states.opml">states.opml</a>

</td>

</tr>

<tr>

<td>

.md

</td>

<td>

The text in the file is passed through a Markdown processor and that text is returned. The type returned is text/html.

</td>

<td>

<a href="http://lucky.wtf/test.md">test.md</a>

</td>

</tr>

<tr>

<td>

.js

</td>

<td>

We run the script, and the return value is returned to the caller, with type of text/html. Here's the <a href="https://gist.github.com/scripting/fd9e6720834958130f0a3d53b9f8dd15">source code</a> for the script in the demo below.

</td>

<td>

<a href="http://lucky.wtf/badass/butt.js">butt.js</a>

</td>

</tr>

</table>

### Port 1339

The first time you run PagePark it will open on port 1339. You can change this by editing prefs.json in the prefs folder. 

This means if you want to access a page on your site, the URL will be of the form:

http://myserver.com:1339/somepage.html

The normal port for HTTP is 80. That would have been the natural default, however a lot of Unix servers require the app to be running in supervisor mode in order for it to open on port 80. You can do this by launching PagePark this way:

sudo node pagepark.js

I made the default 1339 because I wanted it to work "out of the box" for first-time users.

### Mapping port 80 to 1339

Here's a magic incantation that works on Ubuntu that maps requests for port 80 to port 1339.

<pre>sudo iptables -t nat -A PREROUTING -p tcp --dport 80 -j REDIRECT --to-port 1339</pre>

### Example pages

http://morningcoffeenotes.com/ -- simple home page

http://lucky.wtf/ -- images

http://lucky.wtf/test.md -- markdown page

http://lucky.wtf/badass/ -- index file in a sub-directory

http://lucky.wtf/nosuchfile.html -- file not found

http://lucky.wtf/version -- the version of PagePark that's running on the server

http://lucky.wtf/now -- the time on the server

http://lucky.wtf/transcend.opml -- a page written in OPML

http://lucky.wtf/transcend.opml?format=opml -- the OPML source of the page

### config.json files in domains folders

On every request, PagePark looks at the top level of the domain folder for a file named config.json. If it finds it, it reads it and the values in the file control how the request is handled. 

Here's a <a href="https://github.com/scripting/pagePark/blob/master/docs/config.md">docs page</a> that lists the values and what they control.

### JavaScript sample code

I've iterated over the code to try to make it good sample code for JavaScript projects. 

I wanted to make code that could be used for people who are just getting started with Node, to help make the process easier.

There will always be more work to do here. ;-)

### Updates

#### v0.7.31 12/10/19 by DW

You can now override most of the values in the top level config.json file in any of the config.json files in individual domains. It would have worked this way in the beginning if site-level config settings were possible in the beginning. 

Updated the default markdown template to use the same CSS file that GitHub uses. 

#### v0.7.31 12/9/19 by DW

Hidden files and folders

If the file or folder name begins with . do not serve it. 

This feature can be turned off with config.flHiddenFilesCheck.

### v0.7.31 12/4/19 by DW

flProcessScriptFiles defaults false. This is a breaking change. If you have pages that are implemented in JS you will have to switch this on for the site in question. 

Displaying pages of various types with the proper content-type header. 

There was a bug that I fixed, we were calling a local routine named getReturnType with the incorrect parameter. The bug was also present when serving content from S3. 

Display an index file for GitHub directories.

Added a section to the docs for config.json about serving content from GitHub. 

#### v0.7.26 6/8/17 by DW

We look for config.json in the home directory. It contains what used to be in pref/prefs.json. All my other apps work with config.json, I wanted PagePark to be consistent. If we don't find config.json we look for prefs/prefs.json, so that there's no breakage. 

#### v0.7.24 7/6/18 by DW

Added <a href="https://github.com/scripting/pagePark/blob/master/docs/config.md#setting-the-default-content-type-for-a-domain">feature</a> that lets you set the default type for files that don't have extensions. 

#### v0.7.21 6/6/18 by DW

Added <a href="https://github.com/scripting/pagePark/blob/master/docs/config.md#redirect-from-the-file-thats-being-redirected">feature</a> that lets you redirect by modifying the contents of the file.

#### v0.7.19 6/4/18 by DW

Built a bridge to S3 with config.s3ServeFromPath. Now PagePark can serve from S3 locations exactly as if they were in the local file system. I plan to use this as a bridge to publicfolder.io. Updated <a href="https://github.com/scripting/pagePark/blob/master/docs/config.md#serving-content-from-s3">docs</a> for the three S3 configuration options. 

#### v0.7.18 5/12/18 by DW

Continue to have problems with the mime package. Switch to using the daveutils routine instead, does the same thing. One less dependency.

#### v0.7.17 5/10/18 by DW

At some point serving images broke, not sure when. The fix was to not convert the value returned when reading a file to a string. 

#### v0.7.11 2/17/18 by DW

PagePark was failing in httpRespond when filling out logInfo, saying it's undefined. Not clear how this could happen but I added defensive driving to make sure it can't happen, by initializing logInfo at the very start of the routine. 

#### v0.7.9 11/8/17 by DW

A new top-level config.json option, `flUnicasePaths`. 

The feature is <a href="https://github.com/scripting/pagePark/blob/master/docs/config.md#case-sensitive-paths">documented</a>. 

A <a href="http://scripting.com/2017/11/08.html#a102322">blog post</a> about the feature. 

#### v0.7.8 10/3/17 by DW

New features supports logging over WebSockets. To enable, in config.json set <i>flWebsocketEnabled</i> true, and assign a port to the WS server in <i>websocketPort.</i> On every hit, PagePark will send back a JSON structure containing information about the request. 

Note: There's a new dependency, <a href="https://github.com/websockets/ws">nodejs-websocket</a>, so when installing this update you have to do an `npm install` before running pagepark.js.

#### v0.7.7 9/26/17 by DW

Added two config.json options and changed the name of another. 

1. Changed the name of <i>s3Path</i> to <i>fargoS3Path.</i> It was only used in serving Fargo websites, on one of my servers. It's a very specific bit of functionality. Far more specific than the name implies.

2. New meaning for <i>s3Path</i> option. If specified it points to an S3 location that PagePark will serve the site from. It should begin with a slash but not end with one. All requests for the domain are mapped to requests at the S3 location. Additionally, it processes the request through the same filters it uses for local files, so files ending with .md are passed through the Markdown processor and files that end with .opml are processed as outlines. It follows all the rules outlined on <a href="https://github.com/scripting/pagePark/blob/master/docs/config.md">this page</a>.

3. New option: <i>localPath,</i> points to a directory on the local system where files are served from exactly as in #2 above. 

A caveat on this release, I had trouble upgrading one of my servers with this version because of changes to a couple of packages we use. I was able to fix all the problems by installing the latest stable version of Node, as described on <a href="https://nodesource.com/blog/installing-node-js-tutorial-ubuntu/">this page</a>. If you have trouble, re-open <a href="https://github.com/scripting/pagePark/issues/8">this issue</a>, and we can look into it. 

#### v0.7.6 9/16/17 by DW

Two changes:

1. The <a href="https://www.npmjs.com/package/mime">mime package</a> had a breaking change. I fixed it so it no longer breaks PagePark. Thanks to Davis Shaver for the <a href="https://github.com/scripting/pagePark/issues/7">report</a>. 

2. Changed the version number system to use the same x.y.z system used by other Node apps. This is v0.7.6.

#### v0.75 7/17/17 by DW

Check stats every minute instead of every second. This means that servers that take a lot of hits won't be writing the stats file so often. If you run PagePark out of a shared folder this can mean a lot of updating, it's not really protecting against anything critical. 

#### v0.74 7/5/17 by DW

If a request comes in for domain.com/hello and it's a directory, redirect to domain.com/hello/.

#### v0.73 6/17/17 by DW

Replaced <a href="https://www.npmjs.com/package/daveopml">daveopml</a> with <a href="https://www.npmjs.com/package/opmltojs">opmltojs</a>, a new package that builds on the <a href="https://github.com/Leonidas-from-XIV/node-xml2js">xml2js</a> package.

Rebuilt the OPML rendering code to use opmltojs. This allows the &lt;head> info to be transmitted to the rendered page. And the full OPML structure is embedded in the page, meaning the page can render itself without calling back to the server to get missing bits from the OPML.

Added a new pref, <i>flCacheTemplatesLocally,</i> that allows you to say you don't want the OPML and Markdown templates cached. Add it to your prefs.json file. It defaults true, because that was the previous behavior (no breakage). This probably won't matter to people who aren't iterating over template development.

Made the URLs of the OPML and Markdown templates configurable through prefs.json. So if you want to do something nicer than I have, you can, without having to modify pagepark.js.

#### v0.72 6/7/17 by DW

Factored the local <i>utils</i> and <i>opml</i> modules, instead using the new <a href="https://github.com/scripting/utils">daveutils</a> and <a href="https://github.com/scripting/opml">daveopml</a> NPM packages. 

We now look for prefs in config.json in in the same folder as pagepark.js, if it's not present, we look for prefs.json in the prefs folder. All my other server software uses config.json, so PagePark fits in better.

Improved rendering outlines, so that you can put an OPML file anywhere, and refer to it and it will render as an outline. If you add ?format=opml to the URL it will return the XML 

#### v0.71 7/31/16 by DW

The repository now includes <a href="https://github.com/scripting/pagePark/blob/master/prefs/mdTemplate.txt">mdTemplate.txt</a> in the prefs folder. Previously it would download this file from one of my sites the first time it was used. This is a more modern way to distribute it. 

#### v0.70 12/8/15 by DW

It used to be that a file had to exist in order for you to redirect from it in <a href="https://github.com/scripting/pagepark#v067-73015-by-dw">config.redirects</a>. Now it doesn't have to exist. 

#### v0.68 8/31/15 by DW

Small change in error handling when we delegate a request. The previous method would cause PagePark to crash if the app we're trying to delegate to isn't running. Thanks to Dan MacTough for the help fixing this. ;-)

#### v0.67 7/30/15 by DW

New redirect feature for individual pages. 

In config.json for the domain containing the file you want to redirect, create a struct called <i>redirects.</i> The name of each element is the path to a file in the folder, and the value of each is the URL we will redirect to. On each request for that domain, we look in the redirects table to see if the request should be redirected. 

Here's an <a href="https://gist.github.com/scripting/8295f373c61dd9f5ce97">example</a> of the config.json for smallpicture.com. It redirects from an <a href="http://smallpicture.com/outlinerHowto.html">old version</a> of the outliner howto to the newer version. 

#### v0.66 7/19/15 by DW

In prefs.json a new value, legalPathChars, defaults to the empty string. In this string you can specify characters that are legal in paths on your server. We are very conservative in what we will allow in paths, but if you need to use one of the characters that we consider illegal, add it to this string. 

For example, I am redirecting urls from discuss.userland.com to a static archive. It was a Manila site, so it uses a $ in the URLs, a character which PagePark by default considers illegal. Here's an <a href="http://discuss.userland.com/msgReader$18647">example</a>. I set legalPathChars to "$" in prefs.json, and it lets the character through, and it is redirected by the very clever little script that handles redirection for that domain. 

#### v0.65 7/16/15 by DW

In prefs.json a new value, error404File, defaults to prefs/error.html

When there's an error, we read that file and send it back as the text of the 404 response.

The default value is more or less exactly the same text earlier versions of PagePark returned. It's actually in HTML, that's the difference. As opposed to being plain text.

#### v0.64 7/7/15 by DW

I wanted to do a site redirect that was more than just a domain name change. 

If a request came in for: http://archive.scripting.com/2007/12

I wanted it to redirect to: http://scripting.com/2007/12.html

That was accomplished with a new element in config.json: jsSiteRedirect. Its value is a JavaScript expression. You can access the path in the variable parsedUrl.pathname.

This is what config.json in the folder archive.scripting.com looks like. 

<code>{"jsSiteRedirect": "'http://scripting.com' + parsedUrl.pathname + '.html'"}</code>

#### v0.62 6/25/15 by DW

Code cleanup and factoring in <a href="https://github.com/scripting/pagepark/blob/master/lib/opml.js">opml.js</a>.

#### v0.61 6/24/15 by DW

Files with the extension .opml are now  rendered as outlines. 

There's a <a href="http://fargo.io/code/pagepark/defaultopmltemplate.txt">template</a> for outlines, it uses the new <a href="https://github.com/scripting/outlineBrowser">outlineBrowser</a> toolkit. Here's an <a href="http://montauk.scripting.com/test.opml">example</a> of an OPML file rendered through the new template. 

If you want the raw OPML text from the server, you can do it one of two ways:

1. Add ?format=opml at the end of the url. <a href="http://montauk.scripting.com/test.opml?format=opml">Example</a>.

2. If you're making the request in a program, you can add an Accept header with the value: text/x-opml, following a convention established by the OPML Editor. 

As with scripts and markdown files, you can turn the feature off in config.json. 

#### v0.60 5/26/15 by DW

When delegating requests, pass redirects back to the client, don't follow them. This was necessary so that that OAuth dance with Twitter in nodeStorage would work. 

There's a new <a href="http://scripting.com/listings/pagepark.html">structured listing</a> of the source code of PagePark, linked to from the <a href="https://github.com/scripting/pagepark/blob/master/pagepark.js">flat listing</a> (above). This makes it easy to see how the code is organized. It gets pretty deeply nested! The outline view makes that manageable. I use the OPML Editor on a Mac to edit PagePark. 

#### v0.59 5/23/15 by DW

You can now delegate requests to apps running on other ports on your server machine. 

There's a new optional section of the prefs. json file where you specify these mappings. It's called domainMap. It's a set of name-values, where the name is the domain, and the value is the port the requests are mapped to. 

Here's an <a href="https://gist.github.com/scripting/af9df0198db5daa3c982">example</a> of a prefs.json that has a domainMap specified. It maps twitter.radio3.io to the process running on port 5342, twitter.happyfriends.camp to 5338, and any request on judgment.club to the process on port 5351. 

The name matches the end of the HOST header for the request, so a request for judy.judgment.club will map, as will renee.judgment.club and judgment.club.

#### v0.57 5/11/15 by DW

PagePark has pre-defined pages, /now, /version and /status, whose values are returned by PagePark itself. It used to be that they took precedence, so if a site defines pages with those names, the internal ones would be served instead. Now we only serve them if the site didn't define it. 

The <i>urlSiteContents</i> feature now transmits search params. It still will only forward GET calls. This needs to be updated in a future version. 

PagePark now supports wildcards. Suppose you want to serve all the names from mydomain.org with a wildcard. Create a sub-folder of the domains folder with the name *.mydomain.org. If a request comes in for a sub-domain of mydomain.org that doesn't have its own folder, we'll route it through that folder. You can combine this feature with the urlSiteContents feature, or script-implemented pages. 

We also set the <a href="http://stackoverflow.com/questions/19084340/real-life-usage-of-the-x-forwarded-host-header">X-Forwarded-Host</a> and <a href="https://en.wikipedia.org/wiki/X-Forwarded-For">X-Forwarded-For</a> headers on urlSiteContents requests. 

#### v0.56 5/5/15 by DW

New prefs and config values that allow you to disable processing of scripts and Markdown files. By setting the values in prefs.json, you control all domains on the server. And by adding the values to config.json, in the folder the site is served from, you can turn them off selectively by site. I needed to turn off script processing for .js files served from <a href="https://github.com/scripting/river4">River4</a>, to make it possible to serve a full river from PagePark. 

#### v0.55 4/26/15 by DW

With this release you can serve domains whose content is stored elsewhere on the web. 

There's a new optional element of config.json, urlSiteContents. Its value is the URL of a directory on the web that stores the content of the domain. 

As an experiment, I mapped the domain my.this.how to point to a folder in my public Dropbox folder. Here's how I did it.

1. First I created a new sub-folder of my public Dropbox folder called pageParkDemo. I put one file in that folder, index.html.

2. I used the web interface of my domain registrar to point my.this.how at the IP address of my PagePark server.

3. On my PagePark server, I created a sub-folder of the domains folder called my.this.how.

4. I created a <a href="https://gist.github.com/scripting/f06e32acdee685510211">config.json file</a> in that folder, pointing to the pageParkDemo folder in my Dropbox folder.

To test the setup, I just went to <a href="http://my.this.how/">my.this.how</a>, where I saw "this is just a demo".

Just for fun I put a <a href="http://my.this.how/saul.png">picture</a> in that folder, to see if it works.

Obviously it can get a lot more elaborate, and you can store the content anywhere, not just in a Dropbox folder. The key is that the content be accessible over the web. 

#### v0.54 2/18/15 by DW

A new feature for pages implemented as scripts. If the script returns the value <i>undefined</i> PagePark will not return a value to the HTTP client, it assumes that the script will do this. 

To make it possible for a script page to return a value to the client, there's a new built-in function httpReturn. It takes two parameters, the value, a string, and the type, a MIME type. For example you might have a script that makes an HTTP request and returns a value based on the result to the caller.

Here's an <a href="https://gist.github.com/scripting/a3a8232d193ea88e04ba">example script</a> that illustrates.

#### v0.51 1/18/15 by DW

Created utils.js in the lib folder, and require it in pagepark.js.

New feature: If there's a file called config.json in a domain folder, we read it on every request, and values in that file can change the behavior of the server. The first feature allows you to do a whole-site redirect. Useful if you want to have several names map to the same content. Here's an <a href="https://gist.github.com/scripting/27be2d8be40577ad0fdf">example</a> of the config.json file that maps a domain to <a href="http://nodestorage.io/">nodestorage.io</a>.

#### v0.48 1/8/15 by DW

The default port the server boots up on is now 1339. Previously it was 80, which is the standard port for HTTP, but on many OSes this requires PagePark to be running in supervisor mode. I added docs above to explain this. 

Changed package.json so that only *request* and *marked* were listed as dependencies. Apparently the others are included in Node without having to list them. 

Instead of keeping our own MIME type table, we use the Node *mime* package, which is also included as a dependency in the package.json file.

#### v0.47 1/7/15 by DW

Added a package.json file to the repository.

Fixed first-time startup problem creating prefs.json and stats.json. 

Also, we now make sure the *domains* folder exists at startup.

Fixed a problem in handling requests if you specified a different folder for PagePark to serve from. 

### Questions, comments?

Please post a note on the <a href="https://groups.google.com/forum/#!forum/server-snacks">Server Snacks</a> mail list or post an issue <a href="https://github.com/scripting/pagePark/issues">here</a>. 



docs/changes.md
* ### 12/9/19 by DW
   * Hidden files and folders
      * If the file or folder name begins with . do not serve it. 
      * This feature can be turned off with config.flHiddenFilesCheck.
* ### 12/4/19 by DW
   * flProcessScriptFiles defaults false. This is a breaking change. If you have pages that are implemented in JS you will have to switch this on for the site in question. 
   * Displaying pages of various types with the proper content-type header. 
      * There was a bug that I fixed, we were calling a local routine named getReturnType with the incorrect parameter. The bug was also present when serving content from S3. 
   * Display an index file for GitHub directories.
   * Added a section to the docs for config.json about serving content from GitHub. 
* ### 6/8/17 by DW
   * We look for config.json in the home directory. It contains what used to be in pref/prefs.json. All my other apps work with config.json, I wanted PagePark to be consistent. If we don't find config.json we look for prefs/prefs.json, so that there's no breakage. 


lib/opml.js
exports.readOpmlString = readOpmlString;
exports.readOpmlFile = readOpmlFile;
exports.readOpmlUrl = readOpmlUrl;
exports.outlineVisiter = outlineVisiter;

var request = require ("request");
var stream = require ("stream"); //6/23/15 by DW
var opmlParser = require ("opmlparser"); //6/23/15 by DW


var opmlData = { 
	flUseOutlineCache: false,
	outlineCache: new Object (), 
	}

function getBoolean (val) { //12/5/13 by DW
	switch (typeof (val)) {
		case "string":
			if (val.toLowerCase () == "true") {
				return (true);
				}
			break;
		case "boolean":
			return (val);
		case "number":
			if (val == 1) {
				return (true);
				}
			break;
		}
	return (false);
	}
function getNameAtt (theNode) {
	function isAlpha (ch) {
		return (((ch >= 'a') && (ch <= 'z')) || ((ch >= 'A') && (ch <= 'Z')));
		}
	function isNumeric (ch) {
		return ((ch >= '0') && (ch <= '9'));
		}
	function stripMarkup (s) { //5/24/14 by DW
		if ((s === undefined) || (s == null) || (s.length == 0)) {
			return ("");
			}
		return (s.replace (/(<([^>]+)>)/ig, ""));
		}
	function innerCaseName (text) { //8/12/14 by DW
		var s = "", ch, flNextUpper = false;
		text = stripMarkup (text); 
		for (var i = 0; i < text.length; i++) {
			ch = text [i];
			if (isAlpha (ch) || isNumeric (ch)) { 
				if (flNextUpper) {
					ch = ch.toUpperCase ();
					flNextUpper = false;
					}
				else {
					ch = ch.toLowerCase ();
					}
				s += ch;
				}
			else {
				if (ch == ' ') { 
					flNextUpper = true;
					}
				}
			}
		return (s);
		}
	var nameatt = theNode.name;
	if (nameatt === undefined) {
		nameatt = innerCaseName (theNode.text);
		}
	return (nameatt);
	}
function getNodeType (theNode) {
	if (theNode.type == "include") {
		return (theNode.includetype); //this allows include nodes to have types
		}
	else {
		return (theNode.type);
		}
	}
function copyScalars (source, dest) { //8/31/14 by DW
	for (var x in source) { 
		var type, val = source [x];
		if (val instanceof Date) { 
			val = val.toString ();
			}
		type = typeof (val);
		if ((type != "object") && (type != undefined)) {
			dest [x] = val;
			}
		}
	}
function outlineVisiter (theOutline, inlevelcallback, outlevelcallback, nodecallback, visitcompletecallback, flStopAtDocs) {
	function readInclude (theIncludeNode, callback) {
		console.log ("readInclude: url == " + theIncludeNode.url);
		readOpmlUrl (theIncludeNode.url, function (theOutline) {
			if (theOutline === undefined) {
				callback (undefined);
				}
			else {
				expandIncludes (theOutline, function (expandedOutline) {
					callback (expandedOutline); 
					}, flStopAtDocs);
				}
			});
		}
	function typeIsDoc (theNode) {
		if (flStopAtDocs) {
			var type = getNodeType (theNode);
			return ((type !== undefined) && (type != "include") && (type != "link") && (type != "tweet"));
			}
		else {
			return (false);
			}
		}
	function doLevel (head, path, levelcompletecallback) {
		function doOneSub (head, ixsub) {
			if ((head.subs !== undefined) && (ixsub < head.subs.length)) {
				var sub = head.subs [ixsub], subpath = path + getNameAtt (sub);
				if (!getBoolean (sub.iscomment)) { 
					if ((sub.type == "include") && (!typeIsDoc (sub))) {
						nodecallback (sub, subpath);
						readInclude (sub, function (theIncludedOutline) {
							if (theIncludedOutline !== undefined) {
								doLevel (theIncludedOutline, subpath + "/", function () { 
									outlevelcallback ();
									doOneSub (head, ixsub +1);
									});
								}
							else { //6/25/15 by DW -- don't let errors derail us
								doOneSub (head, ixsub +1);
								}
							});
						}
					else {
						if (typeIsDoc (sub)) {
							if (sub.type == "index") {
								subpath += "/";
								}
							nodecallback (sub, subpath);
							doOneSub (head, ixsub +1);
							}
						else {
							nodecallback (sub, subpath);
							if (sub.subs !== undefined) {
								doLevel (sub, subpath + "/", function () { 
									outlevelcallback ();
									doOneSub (head, ixsub +1);
									});
								}
							else {
								doOneSub (head, ixsub +1);
								}
							}
						}
					}
				else {
					doOneSub (head, ixsub +1);
					}
				}
			else {
				levelcompletecallback ();
				}
			}
		inlevelcallback ();
		if (head.type == "include") {
			readInclude (head, function (theIncludedOutline) {
				if (theIncludedOutline !== undefined) {
					doOneSub (theIncludedOutline, 0);
					}
				});
			}
		else {
			doOneSub (head, 0);
			}
		}
	
	if (flStopAtDocs === undefined) { //7/15/15 by DW -- see note at top of routine
		flStopAtDocs = true;
		}
	
	doLevel (theOutline, "", function () {
		outlevelcallback ();
		visitcompletecallback ();
		});
	}
function expandIncludes (theOutline, callback, flStopAtDocs) {
	var theNewOutline = new Object (), lastNewNode = theNewOutline, stack = new Array (), currentOutline;
	function inlevelcallback () {
		stack [stack.length] = currentOutline;
		currentOutline = lastNewNode;
		if (currentOutline.subs === undefined) {
			currentOutline.subs = new Array ();
			}
		}
	function nodecallback (theNode, path) {
		var newNode = new Object ();
		copyScalars (theNode, newNode);
		currentOutline.subs [currentOutline.subs.length] = newNode;
		lastNewNode = newNode;
		}
	function outlevelcallback () {
		currentOutline = stack [stack.length - 1];
		stack.length--; //pop the stack
		}
	
	if (flStopAtDocs === undefined) { //7/15/15 by DW
		flStopAtDocs = true;
		}
	
	outlineVisiter (theOutline, inlevelcallback, outlevelcallback, nodecallback, function () {
		callback (theNewOutline);
		}, flStopAtDocs);
	}
function readOpmlString (s, callback, flExpandIncludes) {
	var opmlparser = new opmlParser ();
	var outlineArray = new Array ();
	var metadata = undefined;
	var flparseerror = false;
	var theStream = new stream.Readable ();
	theStream._read = function noop () {}; 
	theStream.push (s);
	theStream.push (null);
	theStream.pipe (opmlparser);
	
	opmlparser.on ("error", function (error) {
		console.log ("readOpmlString: opml parser error == " + error.message);
		if (callback != undefined) {
			callback (undefined, error);
			}
		flparseerror = true;
		});
	opmlparser.on ("readable", function () {
		var outline;
		while (outline = this.read ()) {
			var ix = Number (outline ["#id"]);
			outlineArray [ix] = outline;
			if (metadata === undefined) {
				metadata = this.meta;
				}
			}
		});
	opmlparser.on ("end", function () {
		if (flparseerror) {
			return;
			}
		var theOutline = new Object ();
		
		//copy elements of the metadata object into the root of the outline
			function copyone (name) {
				var val = metadata [name];
				if ((val !== undefined) && (val != null)) {
					theOutline [name] = val;
					}
				}
			copyone ("title");
			copyone ("datecreated");
			copyone ("datemodified");
			copyone ("ownername");
			copyone ("owneremail");
			copyone ("description");
		
		for (var i = 0; i < outlineArray.length; i++) {
			var obj = outlineArray [i];
			if (obj != null) {
				var idparent = obj ["#parentid"], parent;
				if (idparent == 0) {
					parent = theOutline;
					}
				else {
					parent = outlineArray [idparent];
					}
				if (parent.subs === undefined) {
					parent.subs = new Array ();
					}
				parent.subs [parent.subs.length] = obj;
				delete obj ["#id"];
				delete obj ["#parentid"];
				}
			}
		
		if (flExpandIncludes === undefined) { //7/15/15 by DW
			flExpandIncludes = true;
			}
		if (flExpandIncludes) {
			expandIncludes (theOutline, function (expandedOutline) {
				if (callback != undefined) {
					callback (expandedOutline, undefined);
					}
				}, false);
			}
		else {
			if (callback != undefined) {
				callback (theOutline, undefined);
				}
			}
		});
	}
function readOpmlFile (f, callback) { //6/25/15 by DW 
	fs.readFile (f, function (err, data) {
		if (err) {
			console.log ("readOpmlFile: error reading file " + f + " == " + err.message)
			callback (undefined);
			}
		else {
			readOpmlString (data.toString (), callback);
			}
		});
	}
function readOpmlUrl (urlOutline, callback, flExpandIncludes) { 
	if (flExpandIncludes === undefined) {
		flExpandIncludes = true;
		}
	request (urlOutline, function (err, response, body) {
		if (err !== null) {
			console.log ("readOpmlUrl: error reading file " + urlOutline + " == " + err.message)
			callback (undefined);
			}
		else {
			if (response.statusCode != 200) {
				console.log ("readOpmlUrl: error reading file, statusCode == " + response.statusCode + ", urlOutline == " + urlOutline)
				callback (undefined);
				}
			else {
				readOpmlString (body.toString (), callback, flExpandIncludes);
				}
			}
		});
	}

